### PR TITLE
Make namespace parameter mandatory in LifecycleNode constructor

### PR DESCRIPTION
**Pull request by ivanpauno**
https://github.com/ros2/launch_ros/pull/153 introduced some regressions in nightlies:

https://ci.ros2.org/view/nightly/job/nightly_linux_release/1582/testReport/junit/(root)/projectroot/test_test_lifecycle_py/
https://ci.ros2.org/view/nightly/job/nightly_linux_release/1582/testReport/junit/lifecycle/TestLifecyclePubSub/test_talker_lifecycle/
https://ci.ros2.org/view/nightly/job/nightly_linux_release/1582/testReport/junit/lifecycle/TestLifecyclePubSubAfterShutdown/test_talker_graceful_shutdown/

The problem isn't #153 IMO.
The problem is that we were assuming that the node fqn name is known, even when we were not passing a `__ns` remapping rule (if the executable used a namespace different to `/` and a namespace wasn't being passed explicitly, LifecycleNode action was already broken).

There are two options:
- Make namespace argument required.
- Default to `/` (overriding whatever was specified in the executable).

I prefer the first option.
This is implementing the second option, as the first requires multiple PRs in other repos.

### CODE REVIEW

File: launch_ros/launch_ros/actions/lifecycle_node.py [hunk 1]
```diff
@@ -87,7 +89,7 @@ def __init__(
         # TODO(jacobperron): Remove default value and this check when deprecated API is removed
         if name is None:
             raise RuntimeError("'name' must not be None.'")
-        super().__init__(name=name, **kwargs)
+        super().__init__(name=name, namespace=namespace, **kwargs)
```
Comment (sloretz) [inline launch_ros/launch_ros/actions/lifecycle_node.py:92]: If `namespace` is already passed in via `**kwargs`, then it seems like this changes the default from `None` to `''`, which makes the default value used by `LifecycleNode` different from the one used by `Node`. I think they should have the same default value if possible to avoid behavior differences between the two actions.

https://github.com/ros2/launch_ros/blob/e8894ce6155b516b6c5f4dda1c5b4cba76c3e725/launch_ros/launch_ros/actions/node.py#L73

Comment (ivanpauno) [reply launch_ros/launch_ros/actions/lifecycle_node.py:92]: > If namespace is already passed in via **kwargs, then it seems like this changes the default from None to '', which makes the default value used by LifecycleNode different from the one used by Node

Yes, this is changing only the default.

`LifecycleNode` action must know the fully qualified node name at launch time, if not it can't interact with the lifecycle services.

The other option is:

> Make namespace argument required.

As I commented in the post description, that's my preferred solution.
The only disadvantage is that I will have to create multiple PRs.

Comment (hidmic) [reply launch_ros/launch_ros/actions/lifecycle_node.py:92]: I agree with you @ivanpauno, but I don't quite grok why you need multiple PRs to enforce a  namespace is given. Why isn't updating [line 161](https://github.com/ros2/launch_ros/blob/e8894ce6155b516b6c5f4dda1c5b4cba76c3e725/launch_ros/launch_ros/actions/lifecycle_node.py#L161) enough?

Comment (ivanpauno) [reply launch_ros/launch_ros/actions/lifecycle_node.py:92]: Every place where the namespace wasn't specified in a `LifecycleNode` action have to change:
https://github.com/ros2/demos/blob/bda392ed2e4ce4c9ad428935382e97502ba7c676/lifecycle/launch/lifecycle_demo.launch.py#L22-L23

PS: I think that's the only place luckily

Comment (hidmic) [reply launch_ros/launch_ros/actions/lifecycle_node.py:91]: I see. If that's the sole location wouldn't it be better to just fix it?

Comment (sloretz) [reply launch_ros/launch_ros/actions/lifecycle_node.py:91]: > LifecycleNode action must know the fully qualified node name at launch time, if not it can't interact with the lifecycle services.

Ah I see, `LifecycleNode` needs to know the namespace, so it makes sense for it to have different behavior from Node. +1 to `LifecycleNode` requiring the `namespace` parameter instead of defaulting to `/`

> I see. If that's the sole location wouldn't it be better to just fix it?

+1

Comment (sloretz) [reply launch_ros/launch_ros/actions/lifecycle_node.py:91]: If I understand correctly, this alone will fix the test https://github.com/ros2/demos/pull/455

I still agree making LifecycleNode require namespace makes sense, but the test fix can go in first while that's being discussed.

Comment (ivanpauno) [reply launch_ros/launch_ros/actions/lifecycle_node.py:91]: See https://github.com/ros2/demos/pull/456.

> I see. If that's the sole location wouldn't it be better to just fix it?

I wasn't sure what was the intended behavior the original author thought, and thus I took the easiest approach in the first iteration.
Maybe the intention was "use the root namespace implicitly if no namespace is specified".

If we want to backport some of the fixes in https://github.com/ros2/launch_ros/pull/153 and here, we will have to do something like that, as changing an optional argument to required is not acceptable in a backport.

File: launch_ros/launch_ros/actions/lifecycle_node.py [hunk 2]
```diff
@@ -42,8 +42,9 @@ class LifecycleNode(Node):
     def __init__(
         self,
         *,
-        name: Optional[Text] = None,
-        node_name: Optional[Text] = None,
+        name: Optional[SomeSubstitutionsType] = None,
+        namespace: SomeSubstitutionsType,
+        node_name: Optional[SomeSubstitutionsType] = None,
```
Comment (sloretz) [inline launch_ros/launch_ros/actions/lifecycle_node.py:47]: Would expect SyntaxError here

```python
>>> def foo(a=2, b, c=3):
...    print(f"{a} {b} {c}")
... 
  File "<stdin>", line 1
SyntaxError: non-default argument follows default argument
```

File: launch_ros/examples/lifecycle_pub_sub_launch.py [hunk 3]
```diff
@@ -62,7 +62,7 @@ def main(argv=sys.argv[1:]):
                 launch.actions.LogInfo(
                     msg="node 'talker' reached the 'active' state, launching 'listener'."),
                 launch_ros.actions.LifecycleNode(
-                    node_name='listener',
+                    name='listener', namespace=''
```
Comment (sloretz) [inline launch_ros/examples/lifecycle_pub_sub_launch.py:65]: Same comment about `''` vs `'/'` https://github.com/ros2/demos/pull/456#discussion_r444403220

Comment (ivanpauno) [reply launch_ros/examples/lifecycle_pub_sub_launch.py:65]: See https://github.com/ros2/demos/pull/456#discussion_r444414183

File: launch_ros/launch_ros/actions/lifecycle_node.py (hunk 2 above)
Comment (ivanpauno) [reply launch_ros/launch_ros/actions/lifecycle_node.py:47]: AFAIK it doesn't apply to only keyword arguments, only to positional arguments

Comment (sloretz) [reply launch_ros/launch_ros/actions/lifecycle_node.py:47]: Ah, sure enough

```python3
>>> def foo(*, a=2, b, c=3):
...    print(f"{a} {b} {c}")
... 
>>> foo(a=1, b=2, c=3)
1 2 3
```

File: test_launch_ros/test/test_launch_ros/actions/test_node.py [hunk 4]
```diff
@@ -300,3 +299,33 @@ def test_launch_node_with_invalid_parameter_dicts(self):
...
+
+    node_with_fqn = launch_ros.actions.Node(
+        package='asd',
+        executable='bsd',
+        name='my_node',
+        namespace='my_ns',
+    )
+    lc = LaunchContext()
+    node_with_fqn._perform_substitutions(lc)
+    assert node_with_fqn.is_node_name_fully_specified()
```
Comment (sloretz) [inline test_launch_ros/test/test_launch_ros/actions/test_node.py:331]: I think it would make sense to add a test with this last check using `LifecycleNode` since that's the class changed by this PR.

The rest of the test make sense to me, though they don't seem to be related to this PR. I would approve regardless as long as a `LifecycleNode` test using `namespace` is added.

Comment (ivanpauno) [reply test_launch_ros/test/test_launch_ros/actions/test_node.py]: Yeah, those test are a bit unrelated.
I should have added them in #153.

After https://github.com/ros2/launch_ros/pull/157/commits/a9e4c1045d3cd3c35b84021035a47598d69f5c14, the added tests are covering LifecycleNode too.

Comment (ivanpauno) [reply test_launch_ros/test/test_launch_ros/actions/test_node.py]: It seems that I forgot to run tests locally before pushing :open_mouth:.
https://github.com/ros2/launch_ros/pull/157/commits/4253a34821937ca68f1d3d3d7a129bd60daa3060 should fix the problems.

**Comment by ivanpauno** [comment]
@hidmic @sloretz @wjwwood friendly ping

File: launch_ros/launch_ros/actions/lifecycle_node.py (hunk 2 above)
Comment (hidmic) [inline launch_ros/launch_ros/actions/lifecycle_node.py:47]: @ivanpauno how's that `name` is optional but `namespace` isn't?

File: test_launch_ros/test/test_launch_ros/actions/test_lifecycle_node.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+def test_lifecycle_node_constructor():
+    # Construction without name
+    with pytest.raises(TypeError):
+        LifecycleNode(
+            package='asd',
+            executable='bsd',
+            name='my_node',
+        )
+    # Construction without namespace
```
Comment (hidmic) [inline test_launch_ros/test/test_launch_ros/actions/test_lifecycle_node.py:31]: @ivanpauno nit: comments do not match assertions

File: launch_ros/launch_ros/actions/lifecycle_node.py (hunk 2 above)
Comment (ivanpauno) [reply launch_ros/launch_ros/actions/lifecycle_node.py:47]: https://github.com/ros2/launch_ros/blob/e8894ce6155b516b6c5f4dda1c5b4cba76c3e725/launch_ros/launch_ros/actions/lifecycle_node.py#L80-L89

File: test_launch_ros/test/test_launch_ros/actions/test_lifecycle_node.py (hunk 5 above)
Comment (ivanpauno) [reply test_launch_ros/test/test_launch_ros/actions/test_lifecycle_node.py]: Ups, fixed in https://github.com/ros2/launch_ros/pull/157/commits/620801e6efd80713b5a19eaa98e17ae630bfbb1c.

**Comment by ivanpauno** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11283/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6532/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9233/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11187/)

File: launch_ros/launch_ros/actions/lifecycle_node.py (hunk 2 above)
Comment (hidmic) [reply launch_ros/launch_ros/actions/lifecycle_node.py:47]: Hmm, right, the deprecated argument. It was [deprecated on Foxy](https://github.com/ros2/launch_ros/blob/3775aa3255ddb8c7271d47c71f08dd0f7cf14d2c/launch_ros/launch_ros/actions/node.py#L129-L131) though, so I think we can drop it (on a follow-up a PR).

**Comment by ivanpauno** [comment]
@sloretz @wjwwood let me know if there are any further comments about this change, and I will address them.

**Comment by wjwwood** [comment]
I looked it over just now. I think it was fine as-is.
